### PR TITLE
add UserController, UserBL, and User View

### PR DESCRIPTION
**Pull request by omrishtam**
Home (not logged in): image
Header (logged in): image
Users: image
User Details: image
Edit User: image
Delete User: image

Signed-off-by: omrishtam <omrishtam@gmail.com>

### CODE REVIEW

**Comment by omrishtam** [comment]
[User Controller + View](https://trello.com/c/WuSdL8f1/23-user-controller-view)

File: Controllers/UserController.cs [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        private RecipeBL recipeBL = RecipeBL.GetInstance();
+
+        public object RecipeController { get; private set; }
+
+        // GET: User
+        public async Task<ActionResult> Index()
+        {
+            if (!IsManager())
+            {
+                return Redirect(string.Format("/Error/ErrorPage/{0}", (int)HttpStatusCode.Unauthorized));
```
Comment (shaySirek) [inline Controllers/UserController.cs:29]: If currentUser is null, it's Unauthorized(401) status, but if a regular user (not a manager) is logged in, this it a Forbidden(403) status.
This comment is relevant for all is manager or is logged in checks - the difference between 401 and 403.

we can handle it using enum represents a state(not logged in, logged in as regular, wrong user id, logged in as admin, logged in as system), then implement a function gets id and compare it with current user id and return the state as enum.

so not logged in state will give 401, and wrong user id, or regular user when admin is required, or regular/admin user when system is required will give 403.

File: Views/User/Index.cshtml [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            });
+
+
+
+    </script>
+}
+
+<div class="container">
+
+    <select name="UsersByPrivilege">
```
Comment (shaySirek) [inline Views/User/Index.cshtml:128]: Please use \<ul>\<li>..\</li>\</ul> tags as @RanNissan did at /Views/Ingredient/Index.cshtml.
it looks better...

**Comment by shaySirek** [comment]
@omrishtam could you improve Users page table design:
- there is a line that ends at the middle of the page (above table header).
- make layout look better: you can copy from ingredients table which looks good.

File: Views/User/Details.cshtml [hunk 3]
```diff
@@ -0,0 +1,63 @@
+﻿@using InternetApp2020.Controllers
+@model InternetApp2020.Models.User
+
+
+@{
+    ViewBag.Title = "User details: ";
+}
+
+<div>
+    <dl class="dl-horizontal">
```
Comment (shaySirek) [inline Views/User/Details.cshtml:10]: Please fix fields alignment (aligned to left)

**Comment by omerzamir** [comment]
> @omrishtam could you improve Users page table design:
> 
> * there is a line that ends at the middle of the page (above table header).
> * make layout look better: you can copy from ingredients table which looks good.

I'll push some changes.
@shaySirek

File: Views/User/Details.cshtml [hunk 4]
```diff
@@ -0,0 +1,63 @@
+﻿@using InternetApp2020.Controllers
+@model InternetApp2020.Models.User
+
+
+@{
+    ViewBag.Title = "User details: ";
```
Comment (shaySirek) [inline Views/User/Details.cshtml:6]: ":" should be removed

File: Views/User/Delete.cshtml [hunk 5]
```diff
@@ -0,0 +1,65 @@
+﻿@using InternetApp2020.Controllers
+@model InternetApp2020.Models.User
+
+@{
+    ViewBag.Title = "Delete user: ";
```
Comment (shaySirek) [inline Views/User/Delete.cshtml:5]: ":" should be removed

File: Views/User/Delete.cshtml [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+<h3>Are you sure you want to delete "@Model.UserName"?</h3>
+<div>
+    <dl class="dl-horizontal">
```
Comment (shaySirek) [inline Views/User/Delete.cshtml:10]: Please fix fields alignment (aligned to left)

File: Views/User/Edit.cshtml [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+@{
+    ViewBag.Title = "Edit user";
+}
+
+@using (Html.BeginForm())
+{
+    @Html.AntiForgeryToken()
+
+    <div class="form-horizontal">
```
Comment (shaySirek) [inline Views/User/Edit.cshtml:12]: Please fix fields alignment (aligned to left)

File: Views/User/Index.cshtml [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+              createUserRows(filteredUserData);
+              $(".data-row").click(function() {
+                  window.location = $(this).data("href");
+              });
+            });
+
+
+
+    </script>
+}
```
Comment (omerzamir) [inline Views/User/Index.cshtml:103]: this is horrible we should replace it....

File: Views/Ingredient/Index.cshtml [hunk 9]
```diff
@@ -78,6 +79,7 @@
               //var isUserManager = "UserController.IsManager()";
 
               //if (isUserManager === "True"){
```
[the remark this answers is not in the document]
Comment (shaySirek) [reply Views/Ingredient/Index.cshtml:81]: @RanNissan and I are working on this

File: Controllers/LoginController.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+                                    x.Password == user.Password)).FirstOrDefault();
+
+            if (loggedUser != null)
+            {
+                System.Web.HttpContext.Current.Session["CurrentUser"] = loggedUser;
+                return RedirectToAction("Index", "Home");
+            }
+            ViewBag.ErrorMessage = "Incorrect Username or Password.";
+            return View();
+            //return RedirectToAction("Create", "User");
```
Comment (RanNissan) [inline Controllers/LoginController.cs:32]: Remove comment

File: Controllers/UserController.cs [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+            {
+                return Redirect(string.Format("/Error/ErrorPage/{0}", (int)HttpStatusCode.Forbidden));
+            }
+
+            var users = this.userBL.GetUIUsers();
+
+            return View(users);
+        }
+
+        // GET: User/Details/5
```
Comment (RanNissan) [inline Controllers/UserController.cs:37]: Change the url example for User/Details/:id
there are more like this in this controller.

File: Controllers/UserController.cs [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+
+        // POST: User/Edit/5
+        [HttpPost]
+        [ValidateAntiForgeryToken]
+        public async Task<ActionResult> Edit([Bind(Include = "ID,UserName,FirstName,LastName,Email,IsManager,Password")] User user)
+        {
+            ViewBag.ErrorMessage = "";
+            if (ModelState.IsValid)
+            {
+                User u = (await this.userBL.GetUsers(x => x.UserName == user.UserName)).FirstOrDefault();
```
Comment (RanNissan) [inline Controllers/UserController.cs:125]: better name of variable

File: Models/UserState.cs [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+using System.Linq;
+using System.Web;
+
+namespace InternetApp2020.Models
+{
+    public enum UserState
+    {
+        NotLoggedIn,
+        Regular,
+        RegularForbidden,
```
Comment (RanNissan) [inline Models/UserState.cs:12]: Better name

File: Views/Shared/_Layout.cshtml [hunk 14]
```diff
@@ -25,17 +27,16 @@
...
-                        //if (UserController.IsLoggedIn())
-                        if (true)
+                        if (UserController.IsLoggedIn())
                         {
-                            //User u = UserController.GetLoggedInUser();
-                            <li>@Html.ActionLink("Upload recipe", "Create", "Recipe", new { uid = 1 }, null)</li>
 
-                            //if (UserController.IsManager())
-                            if (true)
+                            User u = UserController.GetLoggedInUser();
```
Comment (RanNissan) [inline Views/Shared/_Layout.cshtml:34]: Better name for variable

File: Views/Shared/_LoginPartial.cshtml [hunk 15]
```diff
@@ -1,27 +1,24 @@
...
-        //if (UserController.IsLoggedIn())
-        if (true)
-        {
-            //User u = UserController.GetLoggedInUser();
+    <ul class="nav navbar-nav navbar-right">
+        @{
+            @Html.AntiForgeryToken()
+            if (UserController.IsLoggedIn())
+            {
+                User u = UserController.GetLoggedInUser();
```
Comment (RanNissan) [inline Views/Shared/_LoginPartial.cshtml:9]: same

File: Controllers/UserController.cs [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+                return Redirect(string.Format("/Error/ErrorPage/{0}", (int)HttpStatusCode.Forbidden));
+            }
+
+            User user = await this.userBL.GetUser(id.Value);
+            if (user == null)
+            {
+                return Redirect(string.Format("/Error/ErrorPage/{0}", (int)HttpStatusCode.NotFound));
+            }
+
+            UserState state = GetPermission(user.ID);
```
Comment (RanNissan) [inline Controllers/UserController.cs:155]: Add UserController. prefix

File: Controllers/LoginController.cs (hunk 10 above)
Comment (omerzamir) [reply Controllers/LoginController.cs]: My bad, forgot to delete it...

**Pull request merged**